File: dart/lib/src/models/tag_summary.dart
//
//  rerout
//  tag_summary.dart
//
//  Official Dart SDK for the Rerout branded-link API.
//
//  Created by Ngonidzashe Mangudya on 20/05/2026.
//  Copyright (c) 2026 Codecraft Solutions
//  Licensed under the MIT License
//  https://codecraftsolutions.co.za
//

import 'package:meta/meta.dart';
import 'package:rerout/src/models/tag.dart';

/// A tag together with the number of live (non-deleted) links it is attached
/// to. Returned by `tags.list()`; the plain [Tag] returned by create/update
/// omits the count.
@immutable
class TagSummary extends Tag {
  /// Creates a [TagSummary].
  const TagSummary({
    required super.id,
    required super.name,
    required super.color,
    required this.linkCount,
  });

  /// Parses a [TagSummary] from the API JSON envelope.
  factory TagSummary.fromJson(Map<String, dynamic> json) => TagSummary(
    id: json['id'] as String,
    name: json['name'] as String,
    color: json['color'] as String,
    linkCount: (json['link_count'] as num?)?.toInt() ?? 0,
  );

  /// Number of live links currently carrying this tag.
  final int linkCount;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TagSummary &&
          id == other.id &&
          name == other.name &&
          color == other.color &&
          linkCount == other.linkCount;

  @override
  int get hashCode => Object.hash(id, name, color, linkCount);

  @override
  String toString() =>
      'TagSummary(id: $id, name: $name, color: $color, linkCount: $linkCount)';
}

File: php/src/Models/TagSummary.php
<?php

declare(strict_types=1);

namespace Rerout\Models;

use InvalidArgumentException;

/**
 * A tag plus its live link count, as returned by `GET /v1/projects/me/tags`.
 *
 * This is the **list-only** shape — `create`/`update` responses return a plain
 * {@see Tag} without `linkCount`. Fields mirror the server-side shape so JSON is
 * parsed without transformation.
 */
final readonly class TagSummary
{
    /**
     * @param string $id        Tag identifier.
     * @param string $name      Human-readable tag name.
     * @param string $color     Display colour (e.g. `teal`).
     * @param int    $linkCount Number of live (non-deleted) links the tag is attached to.
     */
    public function __construct(
        public string $id,
        public string $name,
        public string $color,
        public int $linkCount,
    ) {
    }

    /**
     * Parse a {@see TagSummary} from an API JSON object.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: self::requireString($data, 'id'),
            name: self::requireString($data, 'name'),
            color: self::requireString($data, 'color'),
            linkCount: self::requireInt($data, 'link_count'),
        );
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireString(array $data, string $key): string
    {
        $value = $data[$key] ?? null;
        if (!is_string($value)) {
            throw new InvalidArgumentException("Expected string field '{$key}' in TagSummary payload.");
        }

        return $value;
    }

    /**
     * @param array<string, mixed> $data
     */
    private static function requireInt(array $data, string $key): int
    {
        $value = $data[$key] ?? null;
        if (!is_int($value)) {
            throw new InvalidArgumentException("Expected int field '{$key}' in TagSummary payload.");
        }

        return $value;
    }
}
